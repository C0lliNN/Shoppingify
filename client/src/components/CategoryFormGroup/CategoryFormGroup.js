import PropTypes from 'prop-types';
import React, { useState } from 'react';
import FromGroup from '../UI/FormGroup/FormGroup';
import styled from 'styled-components';
import * as variables from '../../helpers/style-constants';

const ListView = styled.ul`
  border: 1px solid ${variables.COLORS.gray_1};
  box-sizing: border-box;
  padding: 12px;
  list-style: none;
  box-shadow: 0px 2px 12px rgba(0, 0, 0, 0.04);
  border-top-left-radius: ${variables.BORDER_RADIUS_1};
  border-bottom-left-radius: ${variables.BORDER_RADIUS_1};
  height: 145px;
  overflow-y: auto;

  &::-webkit-scrollbar {
    width: 5px;
  }

  &::-webkit-scrollbar-track {
    background: ${variables.COLORS.white};
    border-top-right-radius: ${variables.BORDER_RADIUS_1};
    border-bottom-right-radius: ${variables.BORDER_RADIUS_1};
  }

  &::-webkit-scrollbar-thumb {
    background: ${variables.COLORS.gray_1};
    border-radius: ${variables.BORDER_RADIUS_1};
  }

  &::-webkit-scrollbar-thumb:hover {
    background: ${variables.COLORS.gray_2};
  }

  @media (min-width: ${variables.MD_BREAK_POINT}px) {
    margin-bottom: 100px;
  }
`;

const ListItem = styled.li`
  font-size: ${variables.FONT_SIZE_3};
  line-height: 22px;
  padding: 10px;
  box-sizing: border-box;
  border-radius: ${variables.BORDER_RADIUS_1};
  font-weight: 500;
  font-family: ${variables.FONT_FAMILY};
  color: ${variables.COLORS.gray_3};
  cursor: pointer;
  &:hover,
  &:active {
    background-color: #f2f2f2;
    color: ${variables.COLORS.black_1};
  }
`;

function CategoryFormGroup({ categories, categoryValue, setCategoryValue }) {
  const [showListView, setShowListView] = useState(false);

  function handleInputBlur() {
    setTimeout(() => setShowListView(false), 100);
  }

  function handleInputFocus() {
    setShowListView(true);
  }

  function handleListItemClick(item) {
    setCategoryValue(item);
  }

  const listItems = categories
    .filter((category) => category && category.name)
    .filter(
      (category) =>
        !categoryValue ||
        category.name.toLowerCase().includes(categoryValue.toLowerCase())
    )
    .map((category) => (
      <ListItem
        key={category._id}
        onClick={() => handleListItemClick(category.name)}
      >
        {category.name}
      </ListItem>
    ));

  return (
    <React.Fragment>
      <FromGroup>
        <FromGroup.Label htmlFor="category">Category</FromGroup.Label>
        <FromGroup.Input
          id="category"
          placeholder="Enter a Category"
          onFocus={handleInputFocus}
          onBlur={handleInputBlur}
          value={categoryValue}
          onChange={(event) => setCategoryValue(event.target.value)}
        />
      </FromGroup>
      <ListView style={{ visibility: showListView ? 'inherit' : 'hidden' }}>
        {listItems}
      </ListView>
    </React.Fragment>
  );
}

CategoryFormGroup.propTypes = {
  categories: PropTypes.array.isRequired,
  categoryValue: PropTypes.string.isRequired,
  setCategoryValue: PropTypes.func.isRequired,
};

export default CategoryFormGroup;
