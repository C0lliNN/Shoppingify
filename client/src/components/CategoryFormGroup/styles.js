import styled from 'styled-components';
import * as variables from '../../helpers/style-constants';

export const ListView = styled.ul`
  border: 1px solid ${variables.COLORS.gray_1};
  box-sizing: border-box;
  padding: 12px;
  list-style: none;
  box-shadow: 0px 2px 12px rgba(0, 0, 0, 0.04);
  border-top-left-radius: ${variables.BORDER_RADIUS_1};
  border-bottom-left-radius: ${variables.BORDER_RADIUS_1};
  height: 145px;
  overflow-y: auto;

  &::-webkit-scrollbar {
    width: 5px;
  }

  &::-webkit-scrollbar-track {
    background: ${variables.COLORS.white};
    border-top-right-radius: ${variables.BORDER_RADIUS_1};
    border-bottom-right-radius: ${variables.BORDER_RADIUS_1};
  }

  &::-webkit-scrollbar-thumb {
    background: ${variables.COLORS.gray_1};
    border-radius: ${variables.BORDER_RADIUS_1};
  }

  &::-webkit-scrollbar-thumb:hover {
    background: ${variables.COLORS.gray_2};
  }

  @media (min-width: ${variables.MD_BREAK_POINT}px) {
    margin-bottom: 100px;
  }
`;

export const ListItem = styled.li`
  font-size: ${variables.FONT_SIZE_3};
  line-height: 22px;
  padding: 10px;
  box-sizing: border-box;
  border-radius: ${variables.BORDER_RADIUS_1};
  font-weight: 500;
  font-family: ${variables.FONT_FAMILY};
  color: ${variables.COLORS.gray_3};
  cursor: pointer;
  &:hover,
  &:active {
    background-color: #f2f2f2;
    color: ${variables.COLORS.black_1};
  }
`;