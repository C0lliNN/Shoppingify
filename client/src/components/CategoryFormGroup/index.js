import PropTypes from 'prop-types';
import React, { useState } from 'react';
import FromGroup from '../UI/FormGroup/FormGroup';
import { ListView, ListItem } from './styles';

function CategoryFormGroup({ categories, categoryValue, setCategoryValue }) {
  const [showListView, setShowListView] = useState(false);

  function handleInputBlur() {
    setTimeout(() => setShowListView(false), 100);
  }

  function handleInputFocus() {
    setShowListView(true);
  }

  function handleListItemClick(item) {
    setCategoryValue(item);
  }

  const listItems = categories
    .filter((category) => category && category.name)
    .filter(
      (category) =>
        !categoryValue ||
        category.name.toLowerCase().includes(categoryValue.toLowerCase())
    )
    .map((category) => (
      <ListItem
        key={category._id}
        onClick={() => handleListItemClick(category.name)}
      >
        {category.name}
      </ListItem>
    ));

  return (
    <React.Fragment>
      <FromGroup>
        <FromGroup.Label htmlFor="category">Category</FromGroup.Label>
        <FromGroup.Input
          id="category"
          placeholder="Enter a Category"
          onFocus={handleInputFocus}
          onBlur={handleInputBlur}
          value={categoryValue}
          onChange={(event) => setCategoryValue(event.target.value)}
        />
      </FromGroup>
      {!!listItems.length && (
        <ListView style={{ visibility: showListView ? 'inherit' : 'hidden' }}>
          {listItems}
        </ListView>
      )}
    </React.Fragment>
  );
}

CategoryFormGroup.propTypes = {
  categories: PropTypes.array.isRequired,
  categoryValue: PropTypes.string.isRequired,
  setCategoryValue: PropTypes.func.isRequired,
};

export default CategoryFormGroup;
