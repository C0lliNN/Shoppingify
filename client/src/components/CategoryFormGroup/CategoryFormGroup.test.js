import React from 'react';
import { render } from '@testing-library/react';
import CategoryFormGroup from './CategoryFormGroup';
import 'jest-styled-components';

let categories = null;
let categoryValue = null;
let setCategoryValue = null;

let errorMessage = null;
beforeEach(() => {
  categories = ['Category 1', 'Category 2', 'Category 3', 'Category 4'];
  categoryValue = '';
  setCategoryValue = jest.fn();

  console.error = jest.fn((message) => (errorMessage = message));
});

function exec() {
  return render(
    <CategoryFormGroup
      categories={categories}
      categoryValue={categoryValue}
      setCategoryValue={setCategoryValue}
    />
  );
}

describe('<CategoryFormGroup/>', () => {
  it('should throw an error if categories is not defined', () => {
    categories = null;

    expect(() => exec()).toThrowError();
  });
  it('should throw an erro if categories is not an array', () => {
    categories = 'category';

    expect(() => exec()).toThrowError();
  });
  it('should console.error if categoryValue is not defined', () => {
    categoryValue = null;

    exec();

    expect(console.error).toHaveBeenCalled();
    expect(errorMessage).toMatch(/.*value.*/i);
  });
  it('should console.error if setCategoryValue is not defined', () => {
    setCategoryValue = null;

    exec();

    expect(console.error).toHaveBeenCalled();
    expect(errorMessage).toMatch(/.*setCategory.*required.*/i);
  });
  it('should console.error if setCategoryValue is not a function', () => {
    setCategoryValue = true;

    exec();

    expect(console.error).toHaveBeenCalled();
    expect(errorMessage).toMatch(/.*invalid.*setCategory.*/i);
  });
  it('should render correctly', () => {
    const { container } = exec();

    expect(console.error).not.toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
});
