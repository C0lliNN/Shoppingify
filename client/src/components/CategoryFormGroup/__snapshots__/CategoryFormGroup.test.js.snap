// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryFormGroup/> should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 25px;
}

.c1 {
  display: block;
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
  width: 100%;
  box-sizing: border-box;
  padding: 20px 12px;
  border: 2px solid #BDBDBD;
  border-radius: 12px;
  outline: none;
  font-size: 14px;
  font-family: Quicksand;
  font-weight: 500;
  color: #34333A;
}

.c1:focus {
  border-color: #F9A109;
  box-shadow: 0px 0px 1px rgba(0,0,0,0.3);
}

.c1:focus +label {
  color: #F9A109;
}

.c1::-webkit-input-placeholder {
  color: #BDBDBD;
}

.c1::-moz-placeholder {
  color: #BDBDBD;
}

.c1:-ms-input-placeholder {
  color: #BDBDBD;
}

.c1::placeholder {
  color: #BDBDBD;
}

.c2 {
  display: block;
  width: 100%;
  color: #34333A;
  font-weight: 600;
  margin-bottom: 5px;
  font-size: 14px;
}

.c3 {
  border: 1px solid #E0E0E0;
  box-sizing: border-box;
  padding: 12px;
  list-style: none;
  box-shadow: 0px 2px 12px rgba(0,0,0,0.04);
  border-top-left-radius: 12px;
  border-bottom-left-radius: 12px;
  height: 145px;
  overflow-y: auto;
}

.c3::-webkit-scrollbar {
  width: 5px;
}

.c3::-webkit-scrollbar-track {
  background: #FFFFFF;
  border-top-right-radius: 12px;
  border-bottom-right-radius: 12px;
}

.c3::-webkit-scrollbar-thumb {
  background: #E0E0E0;
  border-radius: 12px;
}

.c3::-webkit-scrollbar-thumb:hover {
  background: #BDBDBD;
}

.c4 {
  font-size: 18px;
  line-height: 22px;
  padding: 10px;
  box-sizing: border-box;
  border-radius: 12px;
  font-weight: 500;
  font-family: Quicksand;
  color: #828282;
  cursor: pointer;
}

.c4:hover,
.c4:active {
  background-color: #f2f2f2;
  color: #000000;
}

@media (min-width:768px) {
  .c3 {
    margin-bottom: 100px;
  }
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      id="category"
      placeholder="Enter a Category"
      value=""
    />
    <label
      class="c2"
      for="category"
    >
      Category
    </label>
  </div>
  <ul
    class="c3"
    style="visibility: hidden;"
  >
    <li
      class="c4"
    />
    <li
      class="c4"
    />
    <li
      class="c4"
    />
    <li
      class="c4"
    />
  </ul>
</div>
`;
